Add grid settings

joergbrech opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

Fixes #1220.

This PR adds the option to set the grid resolution via the settings. Currently, the settings are not preserved and are reset to default values each time TiGL opens. You can set the origin and grid spacing of the grid. The settings are accessible via the cog wheel / Settings menu.

## How Has This Been Tested?

GUI only


## Checklist:

<!--- Our Continuous Integration Workflow will automatically check if all unit tests run without failure -->
<!--- Our Continuous Integration Workflow will automatically check if the code coverage decreases -->
<!--- The following tasks must be performed manually by the contributor and checked by the reviewer -->

<!--- Go over all the following points, and put an `x` in all the boxes in the "Finished" column that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

| Task                                                    | Finished                                                  | Reviewer Approved          |
|---------------------------------------------------------|-----------------------------------------------------------|----------------------------|
| At least one test for the new functionality was added.  | <ul><li>- [ ] yes </li><li>- [x] does not apply</li></ul> | <ul><li>- [ ] OK</li></ul> |
| New classes have been added to the Python interface.    | <ul><li>- [ ] yes </li><li>- [x] does not apply</li></ul> | <ul><li>- [ ] OK</li></ul> |
| The code is properly documented with doxygen docstrings | <ul><li>- [ ] yes </li><li>- [x] does not apply</li></ul> | <ul><li>- [ ] OK</li></ul> |
| Changes are documented at the top of ChangeLog.md       | <ul><li>- [x] yes </li><li>- [ ] does not apply</li></ul> | <ul><li>- [ ] OK</li></ul> |

joergbrech committed:
fix #1220

joergbrech committed:
add Changelog entry for #1220

svengoldberg commented on TIGLCreator/src/TIGLCreatorSettingsDialog.ui:
```diff
@@ -404,6 +443,82 @@
+            <string>Grid Size</string>
+           </property>
+          </widget>
+         </item>
```
Sorry for pickiness, but could you switch the two items to bring them in the "right" (I know that it does not matter here) order...?

joergbrech replied:
This is a computer-written file (QtCreator). IMHO, we can leave it as is. Do you think this is important?

svengoldberg replied:
I see. No, this is of course not really important

svengoldberg commented on TIGLCreator/src/TIGLCreatorSettingsDialog.ui:
```diff
@@ -404,6 +443,82 @@
+         </item>
+         <item row="1" column="1">
+          <widget class="QDoubleSpinBox" name="grid_origin_Y_spinbox">
+           <property name="minimum">
```
Should we also add a `maximum` value here?

joergbrech replied:
Hm, I don't see the maximum value in the ui file, but I see it in QtCreator. Also, I tested it in TiGLCreator and I can't move the grid origin beyond the maximum value.

svengoldberg replied:
Ok 👍

svengoldberg reviewed (changes requested):
I really like this new feature to user-select the grid resolution. Altough I also liked the old somehow automatic resolution, nevertheless I was sometimes annoyed by its "randomness". So, to me, this is an improvement!

Just find some small annotations from my side...

However, one general remark: This change "only" works for the `XY` grid. Tbh, I personally just found out today that there is also the option to choose `XZ` and `YZ`... To me, it looks like that your implementation is not fully compatible right now with a change of the grid.
When I start TiGL and try to choose the `XZ` grid, nothing happens. I need to open the settings and enter `OK`. Then, the change is done. Most likely, this is fixed by a small `emit` of a `SIGNAL`, I think...

joergbrech committed:
Merge branch 'main' into 1220-tigl-creator-provide-more-fine-grained-grid

joergbrech committed:
Apply grid settings immediately when switching grid planes

joergbrech committed:
Apply grid settings immediately when values change in settings dialog

joergbrech committed:
Fix grid spinbox minimum value and fix typo in UI

- Change grid_size_spinbox minimum from 0.0 to 0.001 to prevent invalid grid
- Add validation to reset grid size to 1.0 if <= 0
- Fix typo: 'Orign Y' -> 'Origin Y'

joergbrech committed:
fix minimum value

joergbrech commented on TIGLCreator/src/TIGLCreatorSettingsDialog.ui:
```diff
@@ -404,6 +443,82 @@
+         </item>
+         <item row="0" column="1">
+          <widget class="QDoubleSpinBox" name="grid_origin_X_spinbox">
+           <property name="minimum">
```
See above. I don't see it in the UI file either, but it seems to be in place.

joergbrech committed:
Merge branch 'main' into 1220-tigl-creator-provide-more-fine-grained-grid

svengoldberg committed:
Merge branch 'main' into 1220-tigl-creator-provide-more-fine-grained-grid

svengoldberg commented:
The orientation change is applied immediately now, thanks!.

However, I still have two points:
1. Maybe, I am missing something...
But in Linux, the edited settings (whether it's the changed grid orientation or the changes values) are not preserved when opening TiGL again....

2. When changing the orientation to e.g. YZ, the naming in the settings does not really make sense anymore.... It is still "Origin X" and "Origin Y", which might be a bit confusing... In the YZ case, the box with the name "Origin X" seems to move in Z direction and the "Origin Y" one still in Y direction. However, the order seems a bit weird, when the orientation is called YZ.
Tbh, this is probably a litte annoying to fix and to edit. However, it should be done (potentially in a later PR), because the GUI just does not behave correctly

svengoldberg committed:
Merge branch 'main' into 1220-tigl-creator-provide-more-fine-grained-grid

joergbrech committed:
Merge branch 'main' into 1220-tigl-creator-provide-more-fine-grained-grid

joergbrech committed:
try to persist grid settings via TiGLCreatorSettings/QSettings

joergbrech committed:
properly block signals

joergbrech committed:
persist grid orientation via settings, keep grid origin labels in sync with grid orientation

joergbrech commented:
@svengoldberg I changed the code so that the grid settings are updated on spinbox value change. You don't have to click apply to see the changes. The grid settings are also persisted across sessions via the TiGLCreatorSettings. I changed the setting dialog's labels to be consistent with the current rectangular grid orientation (XY, XZ, YZ). In addition, the grid orientation is also persisted across sessions via the TiGLCreatorSettings.

joergbrech committed:
Merge branch 'main' into 1220-tigl-creator-provide-more-fine-grained-grid

svengoldberg commented:
@joergbrech Cool, I really like the feature and that the settings are stored!

When playing around, it looks to me like the labels for the XZ and YZ case are swapped. Is this right...? Could you just exchange each of the two labels in both these cases?
It appears to me that the movement of the axes is in the respectively other dimension when the `origin` value is changed

joergbrech committed:
proper labels for XZ and YZ grid orientations

joergbrech committed:
Merge branch '1220-tigl-creator-provide-more-fine-grained-grid' of https://github.com/dlr-sc/tigl into 1220-tigl-creator-provide-more-fine-grained-grid

joergbrech commented:
Oh, good catch! Fixed it.

svengoldberg reviewed (approved):
Thanks for the edit!
For me, this is now ready to merge
